enterprise/suppress-user/internal/badgerdb/badgerdb.go

```diff
@@ -0,0 +1,307 @@
+
+	maxSeedWait  time.Duration
+	seederSource func() (io.Reader, error)
+
+	db *badger.DB
+
+	// lock to prevent concurrent access to db during restore
+	restoringLock sync.RWMutex
+	restoring     bool
+	closeOnce     sync.Once
+	closed        chan struct{}
+}
+
+// NewRepository returns a new repository backed by badgerdb.
+func NewRepository(basePath string, log logger.Logger, opts ...Opt) (*Repository, error) {
+	b := &Repository{
+		log:           log,
+		path:          path.Join(basePath, "badgerdbv3"),
+		maxGoroutines: 1,
+		maxSeedWait:   10 * time.Second,
+	}
+	for _, opt := range opts {
+		opt(b)
+	}
+
+	return b, b.start()
+}
+
+// GetToken returns the current token
+func (b *Repository) GetToken() ([]byte, error) {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock() // release the read lock at the end of the operation
+	if b.restoring {
+		return nil, model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return nil, badger.ErrDBClosed
+	}
+
+	var token []byte
+	err := b.db.View(func(txn *badger.Txn) error {
+		item, err := txn.Get([]byte(tokenKey))
+		if err != nil {
+			return fmt.Errorf("could not get token: %w", err)
+		}
+		if err = item.Value(func(val []byte) error {
+			token = val
+			return nil
+		}); err != nil {
+			return fmt.Errorf("could not get token value: %w", err)
+		}
+		return nil
+	})
+	if err != nil && !errors.Is(err, badger.ErrKeyNotFound) {
+		return nil, err
+	}
+	return token, nil
+}
+
+// Suppress returns true if the given user is suppressed, false otherwise
+func (b *Repository) Suppress(workspaceID, userID, sourceID string) (bool, error) {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return false, model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return false, badger.ErrDBClosed
+	}
+
+	keyPrefix := fmt.Sprintf("%s:%s:", workspaceID, userID)
+	err := b.db.View(func(txn *badger.Txn) error {
+		wildcardKey := keyPrefix + model.Wildcard
+		_, err := txn.Get([]byte(wildcardKey))
+		if err == nil {
+			return nil
+		}
+		if !errors.Is(err, badger.ErrKeyNotFound) {
+			return fmt.Errorf("could not get wildcard key %s: %w", wildcardKey, err)
+		}
+		sourceKey := keyPrefix + sourceID
+		if _, err = txn.Get([]byte(sourceKey)); err != nil {
+			return fmt.Errorf("could not get sourceID key %s: %w", sourceKey, err)
+		}
+		return err
+	})
+	if err != nil {
+		if errors.Is(err, badger.ErrKeyNotFound) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// Add adds the given suppressions to the repository
+func (b *Repository) Add(suppressions []model.Suppression, token []byte) error {
+	b.restoringLock.RLock()
+	defer b.restoringLock.RUnlock()
+	if b.restoring {
+		return model.ErrRestoring
+	}
+	if b.db.IsClosed() {
+		return badger.ErrDBClosed
+	}
+	wb := b.db.NewWriteBatch()
+	defer wb.Cancel()
+
+	for i := range suppressions {
+		suppression := suppressions[i]
+		keyPrefix := fmt.Sprintf("%s:%s:", suppression.WorkspaceID, suppression.UserID)
```

Comment:
Given that we create `keyPrefix` in `Suppress()` function in this fashion. Won't it be better to create a function and then use it in both the function, to avoid scope of bug?

Extending this same idea, won't it be better to have a function to create key that we need to search or insert based on given `workspaceID`, `UserID` & `sourceID`?